### ISSUE TITLE
Support for subdivision surfaces using OpenSubdiv

### ISSUE THREAD

**Issue by pouleyKetchoupp** [godotengine/godot-proposals#784]
**Describe the project you are working on:**
Sport game with 3D animation

**Describe the problem or limitation you are having in your project:**
We're using Subdivision Surface Modifier in Blender when creating our models because it allows modelling and animating low-poly models.
https://docs.blender.org/manual/en/latest/modeling/modifiers/generate/subdivision_surface.html

This produces perfectly smooth models in Blender, but once they are exported into Godot, some glitches appear during animation:
(_note: expected results are generated from my work in progress integration of OpenSubdiv_)
| Expected result  | Glitchy result |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/1075032/80860202-7f637a00-8c66-11ea-9bdf-0b30b7152528.png" width="200" />  | <img src="https://user-images.githubusercontent.com/1075032/80860161-46c3a080-8c66-11ea-87d3-c086f6133264.png" width="200" />  |
| <img src="https://user-images.githubusercontent.com/1075032/80860212-8be7d280-8c66-11ea-9677-3ba6002841fc.png" width="200" />  | <img src="https://user-images.githubusercontent.com/1075032/80860216-95713a80-8c66-11ea-8760-b4e5d860a779.png" width="200" />  |
| <img src="https://user-images.githubusercontent.com/1075032/80860257-cf424100-8c66-11ea-89b4-add50a70ea36.gif" width="200" />  | <img src="https://user-images.githubusercontent.com/1075032/80860266-d8cba900-8c66-11ea-90c8-01d1601520e7.gif" width="200" />  |

This is due to the following limitation:
- Export from Blender applies the modifier and generate new vertices with bone weights averaged from the base vertices.
- Godot applies skinning on all vertices from the high-poly model based on the average weights calculated during export, but these weights don't produce a smooth surface when bending.

The solution is to apply the subdivision modifier after the skinning in Godot, but this is not possible at the moment because all we have is the final subdivided mesh.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
This problem can be solved by generating a subdivision mesh in Godot, and interpolate extra vertices after the skinning process to keep surfaces smooth.

We can use OpenSubdiv to help with this process. It's and industry standard, open source and high performance:
http://graphics.pixar.com/opensubdiv

Then we can simply export the low-poly mesh from Blender or other modeling software, apply the subdivision in Godot and get smooth animated meshes:
| Level 0  | Level 1 | Level 2 |
| ------------- | ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/1075032/80860621-395be580-8c69-11ea-83af-21adc8700679.gif" width="200" />  | <img src="https://user-images.githubusercontent.com/1075032/80861276-d1f46480-8c6d-11ea-9223-bf2819cfc678.gif" width="200" />  | <img src="https://user-images.githubusercontent.com/1075032/80860632-48db2e80-8c69-11ea-8c43-f6d7b929e128.gif" width="200" /> |

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
I've made a first basic implementation based on Godot 3.2 for my project:
https://github.com/nekomatata/godot/commit/87e75148281dad03e38f40958bbbddc5ff561765

OpenSubdiv is added as a module to generate subdivision data out of `Mesh` and apply skinning when provided a `Skeleton`.

When the mesh is subdivided, the skinning is done in the OpenSubdiv module directly in order to add a the step to interpolate the extra vertices, so only skinning on CPU is supported in this case.

I've added a property in `MeshInstance` to set a subdivision level, which triggers the generation of the subdivision data and the high-poly rendered mesh.

Here are the next steps for integrating this module into Godot:
- Port to Godot 4.0 / Vulkan
There's not much to do since the whole process is independent from low-level rendering, but updating vertex arrays dynamically is not supported at the moment on master, so that would require fixing before we can integrate it into Godot 4.0. This fix is needed for soft bodies as well.
- Handle normals & UVs
I'm using flat shading in my project so I didn't implement normals and UVs generation, but it shouldn't be too difficult to add.
- Handle blend shapes
Blend shapes are also not supported at the moment. It would require the module to apply blend shapes along with skinning before interpolating the extra vertices from the subdivision.
- Extra optimization
CPU: The skinning process on CPU could be easily optimized a little.
Memory: Share the same vertex array between the different surfaces of a mesh.
Memory: Share subdivision data between instances of the same mesh.

And some extra functionalities would be worth being added in the future:
- Add a subdivision resource
It would allow to optionally generate and store subdivision data offline to trade memory for loading time, and could have a few properties to allow more variation in how to generate the subdivision.
- Support for skinning on GPU
I'm not sure if it's possible and how it would work in the current rendering system, but if the scene shader could access subdivision data we would be able to do the whole skinning/interpolation process on GPU instead of CPU.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
This enhancement can be useful for any artist who wants to use subdivision surfaces. It's too complex to be done with just a few lines of script.

**Is there a reason why this should be core and not an add-on in the asset library?:**
This feature would be best in core, because it needs to take place within the rendering process (after skinning and before rendering) and it requires to be integrated in an optimized way to get good performance.

**Comment by bojidar-bg** [comment]
Question: Is it possible to change the export process, so that the weights of the subdivided vertices are computed correctly?

**Comment by pouleyKetchoupp** [comment]
@bojidar-bg No configuration of weights is going to give you the expected results. Subdivision needs to be applied after the skinning to make sure the surfaces are smoothed correctly.

If you have Blender available, you can easily apply the subdivision surface modifier before or after the armature modifier to see the problem.
![image](https://user-images.githubusercontent.com/1075032/81063620-78dc3900-8ed8-11ea-965e-8581c00a649d.png)
![image](https://user-images.githubusercontent.com/1075032/81063672-94dfda80-8ed8-11ea-83c7-0bd684fd22c6.png)
Here's my test file:
[subdiv-test-new.zip](https://github.com/godotengine/godot-proposals/files/4580545/subdiv-test-new.zip)

In this example, even if you apply the subdivision modifier destructively and re-weight high poly vertices (bending the joint to the right) by hand, you would still have problems when bending the joint to the left.

**Comment by fire** [comment]
Can you describe more about how to implement normals, UVs and blend shapes?

**Comment by pouleyKetchoupp** [comment]
I didn't investigate much in detail how to do the next steps, so after our discussion by dm I'm just leaving some links to find information for anyone interested.

OpenSubdiv documentation:
http://graphics.pixar.com/opensubdiv/docs/api_overview.html

Samples on the OpenSubdiv repo:
https://github.com/PixarAnimationStudios/OpenSubdiv

**Comment by logzero** [comment]
Imho your primary problem is with your vertex weights, not sure if godot is the right place to fix them.

Your low poly mesh in subdiv-test-new has 5 vertex loops from top to bottom and the weights are essentially:
(1, 0) (1, 0) (1, 1) (0, 1) (0, 1)

To get better subdivision/interpolation results change them to;
(1, 0) (0.75, 0.25) (0.5, 0.5) (0.25, 0.75) (0, 1)

You can also check/adjust the weights after baking/applying the subdivision.

**Comment by pouleyKetchoupp** [comment]
@logzero Thanks for your suggestion, I've just tried in Blender to see how it helps. These weight values you're providing allow better interpolation results, but the skinning doesn't look the same. This is not the correct result for joints like character's elbows and knees.

| Original weights  | Smooth weights |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/1075032/81409737-eb9b1d80-913f-11ea-936c-0aa4049d4bb2.png" width="200" />  | <img src="https://user-images.githubusercontent.com/1075032/81409774-ff468400-913f-11ea-9f44-b6fe28f335f7.png" width="200" />  |

Once imported in Godot, it does work without the need for subdivision, but the result is that it deforms the whole shape while bending, so it's not what we're looking for:
<img src="https://user-images.githubusercontent.com/1075032/81410351-d96daf00-9140-11ea-8aaf-de59044b929e.gif" width="200" />

The idea in this proposal is to provide a way for artists to work on the low-poly models, and get the exact same result in Godot. I imagine lots of artists are dealing with workflows where they have to re-adjust weights on the high-poly model. This takes a lot of time, and they can't get the exact same result as with surface subdivision. That's exactly why we've decided to handle subdivision in the game engine, upon request from the artist in our team :)

**Comment by pouleyKetchoupp** [comment]
Also, for anybody interested I'd like to share this GDC presentation about subdivision that @fire mentioned on IRC when we were discussing this proposal:
http://twvideo01.ubm-us.net/o1/vault/GDC2014/Presentations/Brainerd_Wade_Tessellation_in_Call.pdf

It explains why they started to use subdivision in Call of Duty : Ghost to help with the art pipeline, with lots of technical details.

**Comment by fire** [comment]
I'm not able to do much with opensubdiv, but here are some videos of the branch. UVs, materials and normals were added. Skinning is the same as the original.
![GIF 2020-05-08 11-46-48 AM](https://user-images.githubusercontent.com/32321/81481755-31f78780-91e7-11ea-90bf-58cab3c71b88.gif)
![GIF 2020-05-08 9-32-26 PM](https://user-images.githubusercontent.com/32321/81481758-3459e180-91e7-11ea-8e18-8ae6d0384034.gif)

**Comment by DrAlta** [comment]
I'm going to try to hire someone to add render time tessellation as a stretch goal for my MMO Crowd funder. So if anyone how has the skills and the opening in their scheduled to work on this.

[Godot Forum post for the job](https://godotforums.org/discussion/22945/looking-for-someone-to-implement-real-time-tessellation-of-subdivision-surfaces-for-a-stretch-goal/)

**Comment by Ansraer** [comment]
Another major benefit of applying the subdiv modifier in godot are that the shipped 3d models can be significantly smaller.
An exported low poly mesh from blender is 4-6 times smaller than the same mesh after a single subdiv iteration was applied. For a single more detailed character this can quickly add up to hundreds of MBs.

**Comment by Ansraer** [comment]
@fire Could you link your subdiv 3.x branch? I could really use this for the project I am currently working on. If I find some time (and manage to finally fix my 4.0 display scaling PR first) I might even look into upgrading it to 4.x.

**Comment by fire** [comment]
Here is my non-working master https://github.com/V-Sekai/godot/tree/subdivision.

I had some changes, but these are my older branches.

* https://github.com/godot-extended-libraries/godot-fire/tree/feature/3.2/opensubdiv-next
* https://github.com/godot-extended-libraries/godot-fire/tree/feature/3.2/opensubdiv

**Comment by fire** [comment]
Opensubdiv has a catmark operator that works on trimeshes. The blog noted it was a new feature. https://graphics.pixar.com/opensubdiv/docs/release_34.html#triangular-patches-for-loop-subdivision

**Comment by fire** [comment]
https://github.com/tefusion/godot-subdiv 

https://user-images.githubusercontent.com/32321/195954507-ac8eb390-93d9-4386-9196-1aa4596cd583.mp4

- tri mesh -> opensubdiv -> baked into meshinstance3d
- works with skinning (evidence above)
- works with blend shapes (no video)
- requires subdivision bevels and no creases
- works with lod generation

**Comment by Calinou** [comment]
Doing this on import is an interesting idea, but it's not compatible with procedural generation and still causes the file size space in the PCK to increase.

**Comment by fire** [comment]
I didn’t want to ship the opensubdiv gdextension, but your feature of runtime subdiv level has been working for months now.

**Comment by fire** [comment]
No gpu compute shader yet, theres a metal shader we can port.
